### ISSUE TITLE
Fix editor crashes when switching a joint in Leg39 from a CustomJoint to a FreeJoint

### ISSUE THREAD

**Issue by adamkewley** [ComputationalBiomechanicsLab/opensim-creator#263]
Bug report from user:

> I am working on a university project where we need to use opensim creator in order to build a leg model that is sitting on a table (surface) with the knee hanging over the table (surface).
>
> Therefore I tried to edit the already existing Leg39 Model with opensim creator and I followed the same steps as in your tutorial 2 on creating a bouncing block to give the constraints to the Leg39 model on the pelvis (sitting on table) . However the model keeps falling through the contact plane.

I was able to manually reproduce it with the following steps:

1. Open OpenSim Creator v0.1.6
2. Open `leg39.osim` example file
3. Select `hip_r` joint (e.g. through the hierarchy panel)
4. Use Property Editor to change `hip_r` from a `CustomJoint` to a `FreeJoint`
5. (application hard-crashes with no user feedback)

Observations (in `main` branch version of osc):

- It doesn't happen when changing `ground_pelvis` from `CustomJoint` to `FreeJoint`
- `hip_r` completely crashes it
- `knee_r` recognizes that there's a problem (in this case, that there is a constraint in the model that uses a coordinate in `knee_r`). The error is reported to the log and the UI automatically rolls back to the pre-edit version
- `tib_pat_r` behaves the same as `knee_r` (bug recognized and auto-rollback)
- `ankle_r` can be changed from a `CustomJoint` to a `FreeJoint`
- `subtalar_r` can be changed from a `CustomJoint` to a `FreeJoint`
- `mtp_r` can be changed from a `CustomJoint` to a `FreeJoint`

Observations (in released `v0.1.6`):

- They are all a bit flakey: sometimes crashing, sometimes rollbacks, sometimes works
- This would indicate that there's a bad memory access or similar in the algorithm that switches joint types around (the local debug build of `main` might just get luckier?)

**Comment by adamkewley** [comment]
So, deep-diving why this happens:

- `leg39.osim` contains `ConditionalPathPoint`s and other components that indirectly refer to a now-deleted coordinate in the `CustomJoint`
- When the joint is changed, it dirties the model, which prompts OpenSim Creator to ask `OpenSim` to rebuild the model, effectively from scratch
- During a rebuild, OpenSim Creator will rebuild the system from scratch, via this line:

https://github.com/ComputationalBiomechanicsLab/opensim-creator/blob/0.1.6/src/OpenSimBindings/AutoFinalizingModelStatePair.cpp#L190

- The rebuild relies on the fact that `buildSystem` calls `Model::setup`, which calls `Model::finalizeConnections`
- The `Model::finalizeConnections` call works through all components in the model, ensuring that connections between them are made
- Eventually, it works its way towards a component in `leg39.osim` that contains connections to the to-be-deleted coordinate and calls into this code:

https://github.com/opensim-org/opensim-core/blob/4.3/OpenSim/Common/Component.h#L3325

- The code assumes that `connectee`, a cached reference to the connected component, is valid. It isn't--the algorithm deleted the CustomJoint's coordinates during the switch over--and the application then hard-crashes (sometimes, randomly) because the code is using a stale pointer
- The actual return address of where the fault happens is `00007FFE4ED3F90A`:

```x86_64
00007FFE4ED3F8E7  lea         r8,[string " cannot connect to " (07FFE4EF3B838h)]  
00007FFE4ED3F8EE  mov         rdx,rax  
00007FFE4ED3F8F1  lea         rcx,[rbp+130h]  
00007FFE4ED3F8F8  call        std::operator+<char,std::char_traits<char>,std::allocator<char> > (07FFE4EAC2413h)  
00007FFE4ED3F8FD  mov         rbx,rax  
00007FFE4ED3F900  mov         rdx,qword ptr [r12]  
00007FFE4ED3F904  mov         rcx,r12  
00007FFE4ED3F907  call        qword ptr [rdx+10h]  
-----> 00007FFE4ED3F90A  mov         r8,rax  
00007FFE4ED3F90D  mov         rdx,rbx  
00007FFE4ED3F910  lea         rcx,[rbp+150h]  
00007FFE4ED3F917  call        std::operator+<char,std::char_traits<char>,std::allocator<char> > (07FFE4EAB7658h)  
00007FFE4ED3F91C  nop  
00007FFE4ED3F91D  lea         r8,[string " at " (07FFE4EF28FC4h)]  
```

Which, looking at the string append calls either side of it, is when the implementation is calling `comp.getConcreteClassName()`. `Object::getConcreteClassName() const` is a `virtual` function that will (probably) require a vtable lookup. I'm *guessing* the hard-crash fault here is that the indirect call has a junk function pointer (because `comp` is invalid).

So, assuming that analysis is correct I'll need to:

- Write a short reproduction script in `tests/` that automates this fault
- Ensure it actually results in the fault (even if it's a little bit random)
- Add code into the model finalization algorithms that wipes all socket connection caches (if possible), because caching pointers between model versions is a big no-no

**Comment by adamkewley** [comment]
Ok, so I think I now have a test that crashes for the same, or similar, reason that the UI is crashing (the code is very similar).

I now need to implement a downstream fix for this behavior. I have a feeling that it is going to involve breaking the model-reinitialization code a little bit

**Comment by adamkewley** [comment]
The hotfix does the trick, but I now need to go through all uses of sockets in the codebase to figure out where they need to be pre-finalized, because wiping the connections cache has the nasty side-effect of wiping any in-flight connection assignments (of which there may be many)

**Comment by adamkewley** [comment]
Pretty much all model mutations have been ported to the actions layer now.

The next step is to edit the actions layer to ensure that they handle socket connections, finalizing, etc. correctly - along with making the commit messages nicer :>

**Comment by adamkewley** [comment]
This has now been fixed - unless something pops up in v0.2.0 QA

The actual fix here involves clearing all model connections before finalizing the model. The reason why it involved so many changes was because OSC (used to) automatically perform model finalization, connection finalization, etc. and there was no scope for specializations (of not doing it, or doing it in a different order).

There was also other technical debt I've been paying off during this fix. Specifically, the entire back-end now centralizes all model edits in one place, and the model-editing-functions are all designed to manually handle OpenSim-specific stuff, like this bug.

I'll close this ticket and keep the bug in mind when QAing 0.2.0
